File: src/solar/solar-common/util/lang/ProxyUtil.ts
declare interface ObjProxyType {
    vals: any
    getHandlers: ((obj: any, key: any) => any | undefined)[]
    setHandlers: ((obj: any, key: any, val: any) => any | undefined)[]
}

/**
 *  拦截对象的 属性值 访问和设置
 */
export default class ProxyUtil {

    private static objProxys: Map<any, ObjProxyType> = new Map()

    /**
     * 添加一个访问处理器
     * @param obj
     * @param handler
     */
    static addGetHandler(obj: any, handler: ((obj: any, key: any) => any | undefined)) {
        if (targetLanguage != "lua") {
            print("暂未实现")
            return
        }
        ProxyUtil.getObjProxy(obj).getHandlers.push(handler)

    }

    /**
     * 添加一个赋值处理器
     * @param obj
     * @param handler
     */

    static addSetHandler(obj: any, handler: ((obj: any, key: any, val: any) => any | undefined)) {
        if (targetLanguage != "lua") {
            print("暂未实现")
            return
        }
        ProxyUtil.getObjProxy(obj).setHandlers.push(handler)
    }

    /**
     * 添加一个对象的属性值改变的 监听器
     * @param obj
     * @param handler
     */

    static addValueChangeListener(obj: any, key: string, handler: ((oldVal: any, newVal: any) => any | undefined)) {
        if (targetLanguage != "lua") {
            print("暂未实现")
            return
        }
        let lastVal = obj[key];
        ProxyUtil.getObjProxy(obj).setHandlers.push((obj: any, key2: any, val: any) => {
            if (key != key2) {
                return
            }
            if (val != lastVal) {
                handler(lastVal, val)
            }
            lastVal = val;
        });
    }


    private static getObjProxy(obj: any) {
        let objProxy: ObjProxyType = ProxyUtil.objProxys.get(obj);
        if (objProxy != null) {
            return objProxy;
        }
        objProxy = {
            //将监控对象的数据复制到代理对象
            vals: {...obj},
            getHandlers: [],
            setHandlers: [],
        };
        //清空监控对象的数据
        for (let objKey in obj) {
            delete obj[objKey]
        }
        setmetatable(obj, {
            __index: function (this: void, _, k) {
                let result = null;
                for (let handler of objProxy.getHandlers) {
                    result = handler(_, k);
                }
                if (result != null) {
                    return result;
                }
                return objProxy.vals[k];
            } as any,
            __newindex: function (this: void, _, k, v) {
                let result = null;
                for (let handler of objProxy.setHandlers) {
                    result = handler(_, k, v);
                }
                if (result == null) {
                    result = v;
                }
                objProxy.vals[k] = result;
            } as any,
            __pairs: function (this: void) {
                return function (this: void, _, k) {
                    return next(objProxy.vals, k)
                };
            } as any,

            __len: function (this: void) {
                if (objProxy.vals[1] == null) {
                    return 0;
                }
                for (let i = 2; i < 1000000; i++) {
                    if (objProxy.vals[i] == null) {
                        return i;
                    }
                }
            } as any,

        })
        ProxyUtil.objProxys.set(obj, objProxy)
        return objProxy;
    }
}

File: scripts/solar/solar-common/util/lang/ProxyUtil.lua
local ____lualib = require("lualib_bundle")
local __TS__Class = ____lualib.__TS__Class
local __TS__ObjectAssign = ____lualib.__TS__ObjectAssign
local __TS__Delete = ____lualib.__TS__Delete
local Map = ____lualib.Map
local __TS__New = ____lualib.__TS__New
local __TS__SourceMapTraceBack = ____lualib.__TS__SourceMapTraceBack
__TS__SourceMapTraceBack(debug.getinfo(1).short_src, {["10"] = 10,["11"] = 10,["12"] = 10,["14"] = 10,["15"] = 19,["16"] = 20,["17"] = 21,["20"] = 24,["21"] = 24,["22"] = 19,["23"] = 34,["24"] = 35,["25"] = 36,["28"] = 39,["29"] = 39,["30"] = 34,["31"] = 48,["32"] = 49,["33"] = 50,["36"] = 53,["37"] = 54,["38"] = 54,["39"] = 55,["42"] = 58,["43"] = 59,["45"] = 61,["46"] = 54,["47"] = 48,["48"] = 66,["49"] = 67,["50"] = 68,["51"] = 69,["53"] = 71,["54"] = 71,["55"] = 71,["56"] = 71,["57"] = 71,["58"] = 78,["59"] = 79,["61"] = 81,["62"] = 81,["63"] = 81,["64"] = 82,["65"] = 83,["66"] = 84,["67"] = 85,["69"] = 87,["70"] = 88,["72"] = 90,["73"] = 81,["74"] = 92,["75"] = 93,["76"] = 94,["77"] = 95,["79"] = 97,["80"] = 98,["82"] = 100,["83"] = 81,["84"] = 102,["85"] = 103,["86"] = 104,["87"] = 103,["88"] = 81,["89"] = 108,["90"] = 109,["91"] = 110,["94"] = 112,["95"] = 112,["96"] = 113,["97"] = 114,["99"] = 112,["102"] = 81,["103"] = 81,["104"] = 81,["105"] = 120,["106"] = 121,["107"] = 66,["108"] = 12});
local ____exports = {}
____exports.default = __TS__Class()
local ProxyUtil = ____exports.default
ProxyUtil.name = "ProxyUtil"
function ProxyUtil.prototype.____constructor(self)
end
function ProxyUtil.addGetHandler(self, obj, handler)
    if targetLanguage ~= "lua" then
        print("暂未实现")
        return
    end
    local ____table_getHandlers_0 = ____exports.default:getObjProxy(obj).getHandlers
    ____table_getHandlers_0[#____table_getHandlers_0 + 1] = handler
end
function ProxyUtil.addSetHandler(self, obj, handler)
    if targetLanguage ~= "lua" then
        print("暂未实现")
        return
    end
    local ____table_setHandlers_1 = ____exports.default:getObjProxy(obj).setHandlers
    ____table_setHandlers_1[#____table_setHandlers_1 + 1] = handler
end
function ProxyUtil.addValueChangeListener(self, obj, key, handler)
    if targetLanguage ~= "lua" then
        print("暂未实现")
        return
    end
    local lastVal = obj[key]
    local ____table_setHandlers_2 = ____exports.default:getObjProxy(obj).setHandlers
    ____table_setHandlers_2[#____table_setHandlers_2 + 1] = function(____, obj, key2, val)
        if key ~= key2 then
            return
        end
        if val ~= lastVal then
            handler(nil, lastVal, val)
        end
        lastVal = val
    end
end
function ProxyUtil.getObjProxy(self, obj)
    local objProxy = ____exports.default.objProxys:get(obj)
    if objProxy ~= nil then
        return objProxy
    end
    objProxy = {
        vals = __TS__ObjectAssign({}, obj),
        getHandlers = {},
        setHandlers = {}
    }
    for objKey in pairs(obj) do
        __TS__Delete(obj, objKey)
    end
    setmetatable(
        obj,
        {
            __index = function(_, k)
                local result = nil
                for ____, handler in ipairs(objProxy.getHandlers) do
                    result = handler(nil, _, k)
                end
                if result ~= nil then
                    return result
                end
                return objProxy.vals[k]
            end,
            __newindex = function(_, k, v)
                local result = nil
                for ____, handler in ipairs(objProxy.setHandlers) do
                    result = handler(nil, _, k, v)
                end
                if result == nil then
                    result = v
                end
                objProxy.vals[k] = result
            end,
            __pairs = function()
                return function(_, k)
                    return next(objProxy.vals, k)
                end
            end,
            __len = function()
                if objProxy.vals[1] == nil then
                    return 0
                end
                do
                    local i = 2
                    while i < 1000000 do
                        if objProxy.vals[i] == nil then
                            return i
                        end
                        i = i + 1
                    end
                end
            end
        }
    )
    ____exports.default.objProxys:set(obj, objProxy)
    return objProxy
end
ProxyUtil.objProxys = __TS__New(Map)
return ____exports
